fix: highlight calculations race with buffer updates

ollien opened the pull request:
Addresses #1888 (I'm unsure if it's fully fixed, but my preliminary testing shows great improvement. I'll probably run this as a daily driver for a bit to see what I can uncover)

Demo:

https://github.com/vscode-neovim/vscode-neovim/assets/977151/1355c0a8-4e69-499e-ae32-eed44dbf6392



The issue stems from the fact that highlight updates were calculated based on text which may have not yet been committed to the vscode editor. I refactored part of the `HighlightManager` to allow the deferral of this calculation until we can acquire a lock on the editor.

ollien commented:
FWIW, I've been using this for the past couple of days and not seen any strange behavior with highlights :)

justinmk commented:
Very cool. Is there any way to simplify the code? Or to re-use it in other places in the codebase? Or at least delete some old code?

Since you have abstracted `PendingHLUpdates`, it seems like it could get a lot of test coverage simply by writing tests against the `PendingHLUpdates` interface (without involving actualy vscode operations).

xiyaowong commented:
Nice💯

xiyaowong commented on src/highlight_manager.ts:
```diff
@@ -126,3 +146,49 @@ export class HighlightManager implements Disposable {
+        for (const check of checks) {
+            const checkRes = check();
+            if (checkRes) {
+                someCheckTrue = true;
```
```suggestion
                someCheckTrue = true;
                break;
```

justinmk commented:
> there weren't really "unit" tests of things like this, but I'm happy to work some in :)

Ah right. IIRC, the tests don't have access to the direct process, they all go over RPC. So you may need to register a special-purpose vscode command which acts as an interface to the tests.

xiyaowong commented on src/highlight_manager.ts:
```diff
@@ -126,3 +146,49 @@ export class HighlightManager implements Disposable {
+    addGridForceUpdate(grid: number) {
+        this.pushForGrid(grid, () => true);
+    }
+
+    private evaluateChecks(checks: (() => boolean)[]): boolean {
+        let someCheckTrue = false;
+        for (const check of checks) {
+            const checkRes = check();
+            if (checkRes) {
+                someCheckTrue = true;
+            }
+        }
+
+        return someCheckTrue;
+    }
```
```suggestion
```

xiyaowong commented on src/highlight_manager.ts:
```diff
@@ -126,3 +146,49 @@ export class HighlightManager implements Disposable {
+
+    entries(): [number, () => boolean][] {
+        return Array.from(this.pendingGridUpdates.entries()).map(([key, checks]) => {
+            const anyValid = () => this.evaluateChecks(checks);
```
```suggestion
          const anyValid = () => checks.some(check => check());
```

ollien replied:
This actually is subtly incorrect. `.some()` will stop evaluation at the first item that returns `true`, which we don't want here, as we want to process all the elements to calculate the highlights. 

I only know this because it was my first iteration 😅

justinmk replied:
need a (1-line) code comment that explains that

ollien replied:
Sure. And I'll ensure a test validates it too

ollien commented:
_phew_ this PR blew up a bit, but just a summary of the changes I just pushed
1. I added unit tests, which run on every `npm test` invocation. Just for clarity, I renamed some of the existing test scripts to include `integration` explicitly, which blew up the number of files
2. I couldn't find any useful places to reuse this PendingUpdates structure without forcing it a bit, but I did break it out and added tests
3. I also did some minor reshuffling of the `HighlightManager` for clarity :)

justinmk commented on src/test/suite/insert-mode-additions.test.ts:
```diff
@@ -13,7 +13,7 @@ import {
     sendInsertKey,
     sendVSCodeCommand,
-} from "../utils";
+} from "../integrationUtils";
```
looks like unit tests live in `src/test/unit/` , which is good. so probably integ tests should live in `src/test/integ/`.

and if this util module is only for integ tests, should we move it into `src/test/integ/` ?

ollien replied:
Yeah sure. I waffled on this, but I'll get that (and associated docs) cleaned up

justinmk commented on .vscode/launch.json:
```diff
@@ -24,7 +24,7 @@
                 "--disable-extensions",
                 "--extensionDevelopmentPath=${workspaceFolder}",
-                "--extensionTestsPath=${workspaceFolder}/out/test/suite/index"
+                "--extensionTestsPath=${workspaceFolder}/out/test/integration/index"
```
```suggestion
                "--extensionTestsPath=${workspaceFolder}/out/test/integ/index"
```

ollien replied:
Ah, you did say `integ` didn't you. Sec

theol0403 commented:
Thanks for this! I'd keep daily driving the highlights for a bit longer, and then I'm good to merge.

@ollien if you have the time/the will, you are more than welcome to use your newfound knowledge of the highlight provider to make a new PR which continues to simplify things :). It's the piece of this project with the most technical debt atm. Either way, this PR is much appreciated!

ollien commented:
@theol0403 Yeah! I was actually thinking of throwing some unit tests around `HighlightProvider` and working on it. Is there anything in particular you have a particular affliction to (e.g. what carries the most debt)? Or is it just a "big ball of code I'd like to see untangled"?

theol0403 commented:
@ollien Big ball of code, definitely. It's also the source of a large number of issues in this project. Many of those issues are now closed, as @xiyaowong did a lot of work on the highlight provider. However, it can still be simplified.

No pressure at all to fix the following open issues, simplifying the code + adding tests is useful enough!
* https://github.com/vscode-neovim/vscode-neovim/issues/1766
* https://github.com/vscode-neovim/vscode-neovim/issues/1739
* https://github.com/vscode-neovim/vscode-neovim/issues/329#issuecomment-1215278530
* https://github.com/vscode-neovim/vscode-neovim/issues/1367

justinmk commented on CONTRIBUTING.md:
```diff
@@ -45,12 +45,18 @@ You can observe the extension logs via the `vscode-neovim` Output channel or fro
-### Run Tests
+### Run Unit Tests
+
+You can run unit tests independently by doing `npm run test:unit`.
```
```suggestion
Unit tests run in node only (not vscode), and can call code directly (unlike the integration tests).
You can run unit tests independently by doing `npm run test:unit`.
```

ollien commented:
Daily driving is going well :) I'll keep it up. Separately, I've been investigating some of the bugs @theol0403 mentioned this morning, and I think this also fixes #1766, or at least the test case mentioned here https://github.com/vscode-neovim/vscode-neovim/issues/1766#issuecomment-1919864798

On version `1.8.1` if I enter `sample text` on both lines, indent them with `>>` (bottom up), then execute the steps listed here, I get garbled text


https://github.com/vscode-neovim/vscode-neovim/assets/977151/e526d75a-89bf-4f1c-b163-62e2455040a7

However, with this PR, that doesn't seem to happen :tada: 



https://github.com/vscode-neovim/vscode-neovim/assets/977151/a069f214-3a08-4002-85d9-a795c984376e
